### PR TITLE
ISSUE-368 implement TMailContact/autocomplete JMAP method

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

**Comment by Arsnael** [comment]
Pretty much running! 

I just need to fix a little thing... I know we said no sort but we should still do a sort by default right? What should it be? By mail address? Makes sense for a small limit case at least right? Or we don't care and I should just check in my last test I got one item returned regardless of what it is?

**Comment by Arsnael** [comment]
At the end, opted for a default sort on the mail address...

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def validation: Either[Throwable, SMono[Invocation]] = for {
+      limit <- Limit.validateRequestLimit(request.limit)
+    } yield {
+      executeAutocomplete(mailboxSession, request, limit)
+        .map(response => Invocation(
+          methodName = methodName,
+          arguments = Arguments(serializer.serialize(response)),
+          methodCallId = invocation.methodCallId))
+    }
+    validation.fold(SMono.error, res => res)
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala:84]: Limit.validateRequestLimit(request.limit).fold(...) directly ?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  private def executeAutocomplete(session: MailboxSession, request: ContactAutocompleteRequest, limit: Limit.Limit): SMono[ContactAutocompleteResponse] =
+    SFlux.fromPublisher(emailAddressContactSearchEngine.autoComplete(
+        AccountId.fromUsername(session.getUser), request.filter.text.value))
+      .sort(Ordering.by[EmailAddressContact, String](contact => contact.fields.address.asString))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala:90]: IMOwe can remove this and push the sort into the `emailAddressContactSearchEngine`... Cleaner...

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CustomMethod.scala [hunk 3]
```diff
@@ -18,6 +18,7 @@ object CapabilityIdentifier {
   val LINAGORA_FILTER: CapabilityIdentifier = "com:linagora:params:jmap:filter"
   val LINAGORA_PGP: CapabilityIdentifier = "com:linagora:params:jmap:pgp"
   val LINAGORA_LONG_LIVED_TOKEN: CapabilityIdentifier = "com:linagora:params:long:lived:token"
+  val LINAGORA_CONTACT: CapabilityIdentifier = "com:linagora:params:jmap:contact"
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CustomMethod.scala:21]:
```suggestion
  val LINAGORA_CONTACT: CapabilityIdentifier = "com:linagora:params:jmap:contact:autocomplete"
```

**Review by chibenwa — APPROVED** [review]
Small remars

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala (hunk 1 above)
Comment (Arsnael) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala:84]: Tried, can't... does not compile. Saw that we do like this in other parts of the code as well, I guess for the same reason

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala (hunk 2 above)
Comment (Arsnael) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala:90]: Is it? It means you need to think of doing it for every implem?

Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala]: > Is it? It means you need to think of doing it for every implem?

Yes and this means you can push this to the backend! (To ElasticSearch)

Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ContactAutocompleteMethod.scala]: IMO I would even remove this sort compoletly to let ElasticSearch sort things by relevance.

**Comment by vttranlina** [comment]
`com:linagora:params:jmap:contact:autocomplete`
or `com:linagora:params:jmap:autocomplete` ?

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraContactAutocompleteMethodContract.scala [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+
+    val request =
+      s"""{
+         |  "using": ["urn:ietf:params:jmap:core", "com:linagora:params:jmap:autocomplete"],
+         |  "methodCalls": [[
+         |    "TMailContact/autocomplete",
+         |    {
+         |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |      "filter": {"text":"obi"},
+         |      "limit": 1
```
Comment (vttranlina) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraContactAutocompleteMethodContract.scala:652]: Q: 
When the client set `limit` parameter, the response will NOT contain `limit`?
Is it correct?

Comment (Arsnael) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraContactAutocompleteMethodContract.scala:652]: Almost correct. The response will not contain the limit if it is lower than the default set by the server (the one applied then).

If the limit set by client is too high (like 300 while server limit is 256), the used limit has to be specified in the response

Comment (chibenwa) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraContactAutocompleteMethodContract.scala:652]: Yes as the limit is the one set by the client.

The limit is only answered when the value is different from the one supplied by the client.

**Comment by Arsnael** [comment]
> `com:linagora:params:jmap:contact:autocomplete` or `com:linagora:params:jmap:autocomplete` ?

good catch, I read the comment regarding this a bit quick, my bad. should be `com:linagora:params:jmap:contact:autocomplete`

**Review by vttranlina — APPROVED** [review]

**Pull request merged**